arrows/ffmpeg/ffmpeg_video_input.cxx

```diff
@@ -0,0 +1,834 @@
+  }
+
+  AVCodecContext* enc = d->f_format_context->streams[d->f_video_index]->codec;
+
+  // If we have not already converted this frame, try to convert it
+
+  if (!d->current_image_memory && d->f_frame->data[0] != 0)
+  {
+    int width = enc->width;
+    int height = enc->height;
+
+    // If the pixel format is not recognized by then convert the data into RGB_24
+    ffmpeg_video_input::priv::pixel_format format = d->pixel_format_from_ffmpeg(enc->pix_fmt);
+    if (format == ffmpeg_video_input::priv::PIXEL_FORMAT_UNKNOWN)
+    {
+      int size = width * height * 3;
+      if (!d->current_image_memory || size != d->current_image_memory->size())
+      {
+        d->current_image_memory = vital::image_memory_sptr(new vital::image_memory(size));
+      }
+
+      d->f_software_context = sws_getCachedContext(
+        d->f_software_context,
+        width, height, enc->pix_fmt,
+        width, height, AV_PIX_FMT_RGB24,
+        SWS_BILINEAR,
+        NULL, NULL, NULL);
+
+      if (!d->f_software_context)
+      {
+        LOG_ERROR(this->logger(), "Couldn't create conversion context");
+        return nullptr;
+      }
+
+      AVPicture rgb_frame;
+      avpicture_fill(&rgb_frame, (uint8_t*)d->current_image_memory->data(), AV_PIX_FMT_RGB24, width, height);
+
+      int w_step = 1;
+      int h_step = width;
+      int d_step = width*height;
+
+      sws_scale(d->f_software_context,
+        d->f_frame->data, d->f_frame->linesize,
+        0, height,
+        rgb_frame.data, rgb_frame.linesize);
+
+      vital::image image(
+        d->current_image_memory,
+        d->current_image_memory->data(),
+        width, height, 3,
+        3, 3*width, 3*width*height,
+        vital::image_pixel_traits_of<unsigned char>());
+
+      d->current_image = std::make_shared<vital::simple_image_container>(vital::simple_image_container(image));
+    }
+    else
+    {
+      // Test for contiguous memory.  Sometimes FFMPEG uses scanline buffers larger
+      // than the image width.  The extra memory is used in optimized decoding routines.
+      // This leads to a segmented image buffer, not supported by vidl.
+      AVPicture test_frame;
+      avpicture_fill(&test_frame, d->f_frame->data[0], enc->pix_fmt, width, height);
+      if (test_frame.data[1] == d->f_frame->data[1] &&
+        test_frame.data[2] == d->f_frame->data[2] &&
+        test_frame.linesize[0] == d->f_frame->linesize[0] &&
+        test_frame.linesize[1] == d->f_frame->linesize[1] &&
+        test_frame.linesize[2] == d->f_frame->linesize[2])
+      {
+      vital::image image(
+          d->f_frame->data[0],
+          width, height, 3,
+          3, 3 * width, 3 * width*height//,
+          //image_pixel_from_traits_macro from fmt ?
+        );
+      d->current_image = std::make_shared<vital::simple_image_container>(vital::simple_image_container(image));
+      }
+      // Copy the image into contiguous memory.
+      else
+      {
+        if (!d->current_image_memory)
```

Comment:
I would just store `d->current_image` and not `d->current_image_memory`, and always create a new image for each new frame.  The image memory is kind of redundant to store since it's inside the image.